File: android/src/main/java/com/alexmercerind/flutter_media_metadata/MetadataRetriever.java
package com.alexmercerind.flutter_media_metadata;

import android.media.MediaMetadataRetriever;
import android.media.MediaFormat;
import android.media.MediaExtractor;
import java.util.HashMap;
import java.io.File;

public class MetadataRetriever extends MediaMetadataRetriever {
  private String filePath;
  
  public MetadataRetriever() {
    super();
  }

  public void setFilePath(String filePath) {
    this.filePath = filePath;
    setDataSource(filePath);
  }

  public HashMap<String, Object> getMetadata() {
    HashMap<String, Object> metadata = new HashMap<String, Object>();
    
    // Basic track info
    metadata.put("trackName", extractMetadata(METADATA_KEY_TITLE));
    metadata.put("trackArtistNames", extractMetadata(METADATA_KEY_ARTIST));
    metadata.put("albumName", extractMetadata(METADATA_KEY_ALBUM));
    metadata.put("albumArtistName", extractMetadata(METADATA_KEY_ALBUMARTIST));
    
    // Track number parsing (may be in format "1/12")
    String trackNumber = extractMetadata(METADATA_KEY_CD_TRACK_NUMBER);
    try {
      if (trackNumber != null && trackNumber.contains("/")) {
        String[] parts = trackNumber.split("/");
        metadata.put("trackNumber", parts[0].trim());
        metadata.put("albumLength", parts[parts.length - 1].trim());
      } else {
        metadata.put("trackNumber", trackNumber);
        metadata.put("albumLength", null);
      }
    } catch (Exception error) {
      metadata.put("trackNumber", trackNumber);
      metadata.put("albumLength", null);
    }
    
    // Year parsing
    String year = extractMetadata(METADATA_KEY_YEAR);
    String date = extractMetadata(METADATA_KEY_DATE);
    try {
      if (year != null && !year.isEmpty()) {
        metadata.put("year", year.trim());
      } else if (date != null && !date.isEmpty()) {
        // Extract year from date (may be YYYY-MM-DD format)
        metadata.put("year", date.split("-")[0].trim());
      } else {
        metadata.put("year", null);
      }
    } catch (Exception e) {
      metadata.put("year", null);
    }
    
    metadata.put("genre", extractMetadata(METADATA_KEY_GENRE));
    metadata.put("authorName", extractMetadata(METADATA_KEY_AUTHOR));
    metadata.put("writerName", extractMetadata(METADATA_KEY_WRITER));
    metadata.put("discNumber", extractMetadata(METADATA_KEY_DISC_NUMBER));
    metadata.put("mimeType", extractMetadata(METADATA_KEY_MIMETYPE));
    metadata.put("trackDuration", extractMetadata(METADATA_KEY_DURATION));
    metadata.put("bitrate", extractMetadata(METADATA_KEY_BITRATE));
    
    // Note: BPM and Comment are not available via MediaMetadataRetriever
    // They would require parsing ID3 tags directly (e.g., with JAudioTagger library)
    // metadata.put("bpm", null);
    // metadata.put("comment", null);
    
    // Video metadata
    String hasVideo = extractMetadata(METADATA_KEY_HAS_VIDEO);
    if ("yes".equalsIgnoreCase(hasVideo)) {
      metadata.put("width", extractMetadata(METADATA_KEY_VIDEO_WIDTH));
      metadata.put("height", extractMetadata(METADATA_KEY_VIDEO_HEIGHT));
      metadata.put("frameRate", extractMetadata(METADATA_KEY_CAPTURE_FRAMERATE));
      metadata.put("videoCodec", null); // Not directly available via MediaMetadataRetriever
    }
    
    // Audio metadata - try to get from MediaExtractor
    try {
      extractAudioMetadata(metadata);
    } catch (Exception e) {
      // Audio metadata extraction failed, continue without it
    }
    
    // File size
    try {
      File file = new File(filePath);
      if (file.exists()) {
        metadata.put("fileSize", String.valueOf(file.length()));
      }
    } catch (Exception e) {
      // Ignore file size errors
    }
    
    return metadata;
  }
  
  private void extractAudioMetadata(HashMap<String, Object> metadata) {
    MediaExtractor extractor = null;
    try {
      extractor = new MediaExtractor();
      extractor.setDataSource(filePath);
      
      boolean foundAudio = false;
      boolean foundVideo = false;
      
      int trackCount = extractor.getTrackCount();
      for (int i = 0; i < trackCount; i++) {
        MediaFormat format = extractor.getTrackFormat(i);
        String mime = format.getString(MediaFormat.KEY_MIME);
        
        if (!foundAudio && mime != null && mime.startsWith("audio/")) {
          // Audio track found
          foundAudio = true;
          if (format.containsKey(MediaFormat.KEY_CHANNEL_COUNT)) {
            metadata.put("channels", String.valueOf(format.getInteger(MediaFormat.KEY_CHANNEL_COUNT)));
          }
          if (format.containsKey(MediaFormat.KEY_SAMPLE_RATE)) {
            metadata.put("sampleRate", String.valueOf(format.getInteger(MediaFormat.KEY_SAMPLE_RATE)));
          }
          if (format.containsKey(MediaFormat.KEY_BIT_RATE)) {
            metadata.put("audioBitrate", String.valueOf(format.getInteger(MediaFormat.KEY_BIT_RATE)));
          }
          // Extract codec from mime type (e.g., "audio/mp4a-latm" -> "AAC")
          metadata.put("audioCodec", mimeToCodecName(mime));
        }
        
        if (!foundVideo && mime != null && mime.startsWith("video/")) {
          // Video track found
          foundVideo = true;
          if (format.containsKey(MediaFormat.KEY_WIDTH)) {
            metadata.put("width", String.valueOf(format.getInteger(MediaFormat.KEY_WIDTH)));
          }
          if (format.containsKey(MediaFormat.KEY_HEIGHT)) {
            metadata.put("height", String.valueOf(format.getInteger(MediaFormat.KEY_HEIGHT)));
          }
          if (format.containsKey(MediaFormat.KEY_FRAME_RATE)) {
            metadata.put("frameRate", String.valueOf(format.getInteger(MediaFormat.KEY_FRAME_RATE)));
          }
          if (format.containsKey(MediaFormat.KEY_BIT_RATE)) {
            metadata.put("videoBitrate", String.valueOf(format.getInteger(MediaFormat.KEY_BIT_RATE)));
          }
          metadata.put("videoCodec", mimeToCodecName(mime));
        }
        
        // Stop if we found both
        if (foundAudio && foundVideo) {
          break;
        }
      }
    } catch (Exception e) {
      // Extraction failed
    } finally {
      if (extractor != null) {
        extractor.release();
      }
    }
  }
  
  private String mimeToCodecName(String mime) {
    if (mime == null) return null;
    
    // Audio codecs
    if (mime.contains("mp4a") || mime.contains("aac")) return "AAC";
    if (mime.contains("mp3") || mime.contains("mpeg")) return "MP3";
    if (mime.contains("flac")) return "FLAC";
    if (mime.contains("vorbis")) return "Vorbis";
    if (mime.contains("opus")) return "Opus";
    if (mime.contains("wav") || mime.contains("raw")) return "PCM";
    
    // Video codecs
    if (mime.contains("avc") || mime.contains("h264")) return "H.264/AVC";
    if (mime.contains("hevc") || mime.contains("h265")) return "H.265/HEVC";
    if (mime.contains("vp8")) return "VP8";
    if (mime.contains("vp9")) return "VP9";
    if (mime.contains("av01")) return "AV1";
    
    return mime;
  }

  public byte[] getAlbumArt() {
    return getEmbeddedPicture();
  }
  
  /**
   * Get the MIME type of the album art by examining magic bytes.
   * Returns null if no album art or unrecognized format.
   */
  public String getAlbumArtMimeType() {
    byte[] art = getEmbeddedPicture();
    if (art == null || art.length < 4) {
      return null;
    }
    
    // Check magic bytes
    // JPEG: FF D8 FF
    if (art[0] == (byte)0xFF && art[1] == (byte)0xD8 && art[2] == (byte)0xFF) {
      return "image/jpeg";
    }
    // PNG: 89 50 4E 47
    if (art[0] == (byte)0x89 && art[1] == (byte)0x50 && art[2] == (byte)0x4E && art[3] == (byte)0x47) {
      return "image/png";
    }
    // GIF: 47 49 46 38
    if (art[0] == (byte)0x47 && art[1] == (byte)0x49 && art[2] == (byte)0x46 && art[3] == (byte)0x38) {
      return "image/gif";
    }
    // WebP: 52 49 46 46 ... 57 45 42 50
    if (art[0] == (byte)0x52 && art[1] == (byte)0x49 && art[2] == (byte)0x46 && art[3] == (byte)0x46 && art.length > 11) {
      if (art[8] == (byte)0x57 && art[9] == (byte)0x45 && art[10] == (byte)0x42 && art[11] == (byte)0x50) {
        return "image/webp";
      }
    }
    // BMP: 42 4D
    if (art[0] == (byte)0x42 && art[1] == (byte)0x4D) {
      return "image/bmp";
    }
    
    return null;
  }
}

File: ios/Classes/MetadataRetriever.swift
import AVFoundation
import Foundation

#if os(macOS)
  import CoreServices
#elseif os(iOS)
  import MobileCoreServices
#endif

protocol MetadataRetrieverProtocol {
  func getTrackName() -> String?
  func getArtistNames() -> String?
  func getAlbumName() -> String?
  func getAlbumArtistName() -> String?
  func getTrackNumber() -> String?
  func getAlbumLength() -> String?
  func getYear() -> String?
  func getGenre() -> String?
  func getAuthorName() -> String?
  func getWriterName() -> String?
  func getDiscNumber() -> String?
  func getBpm() -> String?
  func getComment() -> String?
  func getAlbumArt() -> Data?
}

public class MetadataRetriever {
  let filePath: String
  let url: URL
  let asset: AVAsset
  let preferredRetriever: MetadataRetrieverProtocol?

  init(_ filePath: String) {
    self.filePath = filePath
    self.url = URL(fileURLWithPath: self.filePath)

    let asset = AVAsset(url: url)
    let preferredRetriever: MetadataRetrieverProtocol? = {
      if asset.availableMetadataFormats.contains(.id3Metadata) {
        return Id3MetadataRetriever(asset.metadata(forFormat: .id3Metadata))
      }

      if asset.availableMetadataFormats.contains(.iTunesMetadata) {
        return ItunesMetadataRetriever(asset.metadata(forFormat: .iTunesMetadata))
      }

      return nil
    }()

    self.asset = asset
    self.preferredRetriever = preferredRetriever
  }

  public func getMetadata() -> [String: Any] {
    var metadata: [String: Any] = [:]

    // Basic track info
    metadata["trackName"] = getTrackName()
    metadata["trackArtistNames"] = getArtistNames()
    metadata["albumName"] = getAlbumName()
    metadata["albumArtistName"] = getAlbumArtistName()
    metadata["trackNumber"] = getTrackNumber()
    metadata["albumLength"] = getAlbumLength()
    metadata["year"] = getYear()
    metadata["genre"] = getGenre()
    metadata["authorName"] = getAuthorName()
    metadata["writerName"] = getWriterName()
    metadata["discNumber"] = getDiscNumber()
    metadata["mimeType"] = getMimeType()
    metadata["trackDuration"] = getDuration()
    metadata["bitrate"] = getBitrate()
    metadata["bpm"] = getBpm()
    metadata["comment"] = getComment()
    
    // File size
    metadata["fileSize"] = getFileSize()
    
    // Audio track metadata
    if let audioMetadata = getAudioTrackMetadata() {
      for (key, value) in audioMetadata {
        metadata[key] = value
      }
    }
    
    // Video track metadata
    if let videoMetadata = getVideoTrackMetadata() {
      for (key, value) in videoMetadata {
        metadata[key] = value
      }
    }
    
    // Format info
    metadata["format"] = getFormat()
    
    return metadata
  }

  private func getTrackName() -> String? {
    return preferredRetriever?.getTrackName()
  }

  private func getArtistNames() -> String? {
    return preferredRetriever?.getArtistNames()
  }

  private func getAlbumName() -> String? {
    return preferredRetriever?.getAlbumName()
  }

  private func getAlbumArtistName() -> String? {
    return preferredRetriever?.getAlbumArtistName()
  }

  private func getTrackNumber() -> String? {
    return preferredRetriever?.getTrackNumber()
  }

  private func getAlbumLength() -> String? {
    return preferredRetriever?.getAlbumLength()
  }

  private func getYear() -> String? {
    return preferredRetriever?.getYear()
  }

  private func getGenre() -> String? {
    return preferredRetriever?.getGenre()
  }

  private func getAuthorName() -> String? {
    return preferredRetriever?.getAuthorName()
  }

  private func getWriterName() -> String? {
    return preferredRetriever?.getWriterName()
  }

  private func getDiscNumber() -> String? {
    return preferredRetriever?.getDiscNumber()
  }
  
  private func getBpm() -> String? {
    return preferredRetriever?.getBpm()
  }
  
  private func getComment() -> String? {
    return preferredRetriever?.getComment()
  }

  private func getDuration() -> String {
    let milliseconds = Int(Float64(asset.duration.value * 1000) / Float64(asset.duration.timescale))
    return String(milliseconds)
  }

  private func getBitrate() -> String? {
    // NOTE: AVAssetTrack:estimatedDataRate returns 0.0 if file is mp3
    var audioFileRef: ExtAudioFileRef?
    let openFileResult = ExtAudioFileOpenURL(url as CFURL, &audioFileRef)
    guard let audioFileRef = audioFileRef, openFileResult == 0 else {
      return nil
    }

    var audioFileID: AudioFileID?
    var propertyDataSize = UInt32(MemoryLayout<AudioFileID>.size)
    let getPropertyResult = ExtAudioFileGetProperty(
      audioFileRef, kExtAudioFileProperty_AudioFile, &propertyDataSize, &audioFileID)
    guard let audioFileID = audioFileID, getPropertyResult == 0 else {
      return nil
    }

    var bitRate: UInt32 = 0
    var bitRateSize = UInt32(MemoryLayout.size(ofValue: bitRate))
    let getBitRateResult = AudioFileGetProperty(
      audioFileID, kAudioFilePropertyBitRate, &bitRateSize, &bitRate)
    guard getBitRateResult == 0 else {
      return nil
    }
    return String(bitRate)
  }

  private func getMimeType() -> String? {
    let pathExtension = self.url.pathExtension
    guard
      let identifier = UTTypeCreatePreferredIdentifierForTag(
        kUTTagClassFilenameExtension, pathExtension as NSString, nil)?.takeRetainedValue()
    else {
      return nil
    }

    return UTTypeCopyPreferredTagWithClass(identifier, kUTTagClassMIMEType)?.takeRetainedValue()
      as? String
  }
  
  private func getFileSize() -> String? {
    do {
      let attributes = try FileManager.default.attributesOfItem(atPath: filePath)
      if let fileSize = attributes[.size] as? Int64 {
        return String(fileSize)
      }
    } catch {
      // Ignore errors
    }
    return nil
  }
  
  private func getFormat() -> String? {
    let pathExtension = self.url.pathExtension.uppercased()
    switch pathExtension {
    case "MP3":
      return "MPEG Audio"
    case "M4A", "AAC":
      return "AAC"
    case "FLAC":
      return "FLAC"
    case "WAV":
      return "WAV"
    case "OGG":
      return "Ogg"
    case "MP4", "M4V":
      return "MPEG-4"
    case "MOV":
      return "QuickTime"
    case "AVI":
      return "AVI"
    case "MKV":
      return "Matroska"
    default:
      return pathExtension
    }
  }
  
  private func getAudioTrackMetadata() -> [String: Any]? {
    guard let audioTrack = asset.tracks(withMediaType: .audio).first else {
      return nil
    }
    
    var metadata: [String: Any] = [:]
    
    // Get format descriptions
    if let formatDescriptions = audioTrack.formatDescriptions as? [CMFormatDescription],
       let formatDescription = formatDescriptions.first {
      
      // Get audio stream basic description
      if let asbd = CMAudioFormatDescriptionGetStreamBasicDescription(formatDescription)?.pointee {
        metadata["sampleRate"] = String(Int(asbd.mSampleRate))
        metadata["channels"] = String(asbd.mChannelsPerFrame)
        
        if asbd.mBitsPerChannel > 0 {
          metadata["bitDepth"] = String(asbd.mBitsPerChannel)
        }
      }
      
      // Get codec
      let mediaSubType = CMFormatDescriptionGetMediaSubType(formatDescription)
      metadata["audioCodec"] = fourCCToCodecName(mediaSubType)
    }
    
    // Estimated bitrate
    let estimatedBitrate = audioTrack.estimatedDataRate
    if estimatedBitrate > 0 {
      metadata["audioBitrate"] = String(Int(estimatedBitrate))
    }
    
    return metadata
  }
  
  private func getVideoTrackMetadata() -> [String: Any]? {
    guard let videoTrack = asset.tracks(withMediaType: .video).first else {
      return nil
    }
    
    var metadata: [String: Any] = [:]
    
    // Dimensions
    let size = videoTrack.naturalSize
    metadata["width"] = String(Int(size.width))
    metadata["height"] = String(Int(size.height))
    
    // Frame rate
    let frameRate = videoTrack.nominalFrameRate
    if frameRate > 0 {
      metadata["frameRate"] = String(format: "%.3f", frameRate)
    }
    
    // Estimated bitrate
    let estimatedBitrate = videoTrack.estimatedDataRate
    if estimatedBitrate > 0 {
      metadata["videoBitrate"] = String(Int(estimatedBitrate))
    }
    
    // Get format descriptions for codec info
    if let formatDescriptions = videoTrack.formatDescriptions as? [CMFormatDescription],
       let formatDescription = formatDescriptions.first {
      let mediaSubType = CMFormatDescriptionGetMediaSubType(formatDescription)
      metadata["videoCodec"] = fourCCToCodecName(mediaSubType)
    }
    
    return metadata
  }
  
  private func fourCCToCodecName(_ fourCC: FourCharCode) -> String {
    switch fourCC {
    // Audio codecs
    case kAudioFormatMPEGLayer3:
      return "MP3"
    case kAudioFormatMPEG4AAC, kAudioFormatMPEG4AAC_HE, kAudioFormatMPEG4AAC_HE_V2:
      return "AAC"
    case kAudioFormatAppleLossless:
      return "ALAC"
    case kAudioFormatFLAC:
      return "FLAC"
    case kAudioFormatLinearPCM:
      return "PCM"
    case kAudioFormatOpus:
      return "Opus"
    // Video codecs
    case kCMVideoCodecType_H264:
      return "H.264/AVC"
    case kCMVideoCodecType_HEVC:
      return "H.265/HEVC"
    case kCMVideoCodecType_VP9:
      return "VP9"
    case kCMVideoCodecType_AV1:
      return "AV1"
    default:
      // Convert FourCC to string
      let chars = [
        Character(UnicodeScalar((fourCC >> 24) & 0xFF)!),
        Character(UnicodeScalar((fourCC >> 16) & 0xFF)!),
        Character(UnicodeScalar((fourCC >> 8) & 0xFF)!),
        Character(UnicodeScalar(fourCC & 0xFF)!)
      ]
      return String(chars).trimmingCharacters(in: .whitespaces)
    }
  }

  public func getAlbumArt() -> Data? {
    return preferredRetriever?.getAlbumArt()
  }
  
  /// Get the MIME type of the album art by examining magic bytes.
  public func getAlbumArtMimeType() -> String? {
    guard let data = getAlbumArt(), data.count >= 4 else {
      return nil
    }
    
    let bytes = [UInt8](data.prefix(12))
    
    // JPEG: FF D8 FF
    if bytes[0] == 0xFF && bytes[1] == 0xD8 && bytes[2] == 0xFF {
      return "image/jpeg"
    }
    // PNG: 89 50 4E 47
    if bytes[0] == 0x89 && bytes[1] == 0x50 && bytes[2] == 0x4E && bytes[3] == 0x47 {
      return "image/png"
    }
    // GIF: 47 49 46 38
    if bytes[0] == 0x47 && bytes[1] == 0x49 && bytes[2] == 0x46 && bytes[3] == 0x38 {
      return "image/gif"
    }
    // WebP: 52 49 46 46 ... 57 45 42 50
    if bytes.count >= 12 && bytes[0] == 0x52 && bytes[1] == 0x49 && bytes[2] == 0x46 && bytes[3] == 0x46 {
      if bytes[8] == 0x57 && bytes[9] == 0x45 && bytes[10] == 0x42 && bytes[11] == 0x50 {
        return "image/webp"
      }
    }
    // BMP: 42 4D
    if bytes[0] == 0x42 && bytes[1] == 0x4D {
      return "image/bmp"
    }
    
    return nil
  }
}
